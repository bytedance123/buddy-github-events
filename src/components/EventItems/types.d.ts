export interface Commit {
  sha: string
  author: Author
  message: string
  distinct: boolean
  url: string
}

interface Author {
  email: string
  name: string
}
