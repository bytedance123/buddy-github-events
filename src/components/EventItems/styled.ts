import styled from 'styled-components'

export const Wrapper = styled.div`
  display: flex;
  padding: 8px 0px;
  margin: 8px 16px;
  border-bottom: 1px solid #e1e4e8;
`

export const Avatar = styled.img`
  border: 1px solid #e1e4e8;
  border-radius: 100%;
`

export const Descript = styled.p`
  margin: 8px;

  a {
    color: #24292e;
    text-decoration: none;
    font-weight: 600;
  }

  a:hover {
    color: #0366d6;
  }
`
