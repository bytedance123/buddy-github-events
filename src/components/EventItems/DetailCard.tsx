import React from 'react'
import styled, { css } from 'styled-components'
import ReactMarkdown from 'react-markdown'
import { GithubApi } from 'parse-github-event/lib/types'

import Card, { INTENTS } from '../Card'
import { spLayout } from '../Layout'
import { Tag } from '../Tag'

const monospacedFonts = css`
  font-family: SFMono-Regular, Consolas, Liberation Mono, Menlo, Courier, monospace;
`

const DetailCardWrapper = styled(Card)`
  overflow: hidden;
  word-break: break-word;

  h5 {
    font-weight: 600;
    font-size: 16px;
    margin: 8px 0;

    /* issue number */
    span {
      margin: 0 8px;
      color: #6a737d;
    }
  }

  a {
    text-decoration: none;
    color: #24292e;
    &:hover {
      color: #0366d6;
    }
  }

  details {
    ${spLayout(`
      display: none;
    `)}
    img {
      max-width: 500px;
    }
    pre {
      ${monospacedFonts}
      background-color: #f6f8fa;
      border-radius: 3px;
      font-size: 85%;
      line-height: 1.45;
      max-width: 500px;
      overflow: auto;
      padding: 16px;
    }
  }
`

const CommitMessage = styled.p`
  ${monospacedFonts}

  font-size: 14px;
  margin: unset;
  margin: 4px;
`

const GITHUB_URL = 'https://github.com'

const DetailCard = ({ event }: { event: GithubApi.GithubEvent }) => {
  // https://developer.github.com/v3/activity/events/types/
  switch (event.type) {
    case 'IssuesEvent':
    case 'IssueCommentEvent':
      return (
        <DetailCardWrapper intent={INTENTS.NONE}>
          <h5>
            <a href={event.payload.issue.html_url} rel="noopener noreferrer" target="_blank">
              {event.payload.issue.title}
            </a>
            <span>#{event.payload.issue.number}</span>
          </h5>
          {event.payload.issue.body && (
            <details>
              <ReactMarkdown source={event.payload.issue.body} />
            </details>
          )}
        </DetailCardWrapper>
      )
    case 'PullRequestEvent':
      return (
        <DetailCardWrapper intent={INTENTS.NONE}>
          <h5>
            <a href={event.payload.pull_request.html_url} rel="noopener noreferrer" target="_blank">
              {event.payload.pull_request.title}
            </a>
            <span>#{event.payload.pull_request.number}</span>
          </h5>
          {event.payload.pull_request.body && (
            <details>
              <ReactMarkdown source={event.payload.pull_request.body} />
            </details>
          )}
        </DetailCardWrapper>
      )
    case 'PushEvent':
      return event.payload.commits.length > 0 ? (
        <DetailCardWrapper intent={INTENTS.NONE}>
          {event.payload.commits.map((
            commit: any, // TODO fix type https://developer.github.com/v3/repos/commits/
          ) => (
            <CommitMessage key={commit.sha}>
              <a
                href={`${GITHUB_URL}/${event.repo.name}/commit/${commit.sha}`}
                rel="noopener noreferrer"
                target="_blank"
              >
                <Tag round>{commit.sha.slice(0, 7)}</Tag>
              </a>
              {commit.message.split('\n')[0]}
            </CommitMessage>
          ))}
        </DetailCardWrapper>
      ) : null
    default:
      return null
  }
}

export default DetailCard
