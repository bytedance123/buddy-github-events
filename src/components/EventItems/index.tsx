import React, { Fragment } from 'react'
import { withTranslation, WithTranslation } from 'react-i18next'
import { Event, parse } from 'parse-github-event'
import TimeAgo from 'react-timeago'

import { Wrapper, Avatar, Descript } from './styled'
import Card, { INTENTS } from '../Card'

const GITHUB_DOMAIN = 'https://github.com'

const EventItem = (event: Event) => {
  const {
    actor: { login, avatar_url },
  } = event

  const actorUrl = login.endsWith('[bot]') // handle bot
    ? `${GITHUB_DOMAIN}/apps/${login.slice(0, -'[bot]'.length)}`
    : `${GITHUB_DOMAIN}/${login}`

  const eventParsed = parse(event)

  if (!eventParsed) {
    console.error('github event parsed fail!', event)
    return null // TODO: ErrorCompents
  }

  const templateDate: any = eventParsed.data // TODO fix type
  if (eventParsed.data.hasOwnProperty('repository')) {
    templateDate['repository'] = (
      <a
        key={'repo'}
        href={`${GITHUB_DOMAIN}/${templateDate['repository']}`}
        target="_blank"
        rel="noopener noreferrer"
      >
        {templateDate['repository']}
      </a>
    )
  }
  if (eventParsed.data.hasOwnProperty('member')) {
    templateDate['member'] = (
      <a
        key={'member'}
        href={templateDate['member']['html_url']}
        target="_blank"
        rel="noopener noreferrer"
      >
        {templateDate['member']['login']}
      </a>
    )
  }

  const descriptArray: React.ReactNode[] = [
    <Fragment key={'login'}>
      <a href={actorUrl} target="_blank" rel="noopener noreferrer">
        {login}
      </a>
      &nbsp;
    </Fragment>,
  ]

  // parse template
  const regex = /{{(.*?)}}/g
  const { text } = eventParsed
  let startIndex = 0
  let matchResult
  while ((matchResult = regex.exec(text))) {
    descriptArray.push(text.slice(startIndex, matchResult.index))
    descriptArray.push((eventParsed.data as any)[matchResult[1]]) // TODO: fix type
    startIndex = regex.lastIndex
  }

  return (
    <Wrapper>
      <a href={actorUrl} target="_blank" rel="noopener noreferrer">
        <Avatar src={avatar_url} width={'32'} height={'32'} alt={`@${login}`} />
      </a>
      <Descript>{descriptArray}</Descript>
      <TimeAgo date={new Date(event.created_at)} />
    </Wrapper>
  )
}

export interface EventItemsProps {
  events?: Event[]
}

const EventItems = ({ events, t }: EventItemsProps & WithTranslation) => {
  if (!events) {
    return null
  }
  if (events.length === 0) {
    return <Card intent={INTENTS.WARNING}>{t('No recent activity')}</Card>
  }
  return (
    <div>
      {events
        .filter((e, index, self) => index === self.findIndex(t => t.id === e.id)) // remove id duplicates
        .map(event => (
          <EventItem key={event.id} {...event} />
        ))}
    </div>
  )
}

export default withTranslation()(EventItems)
