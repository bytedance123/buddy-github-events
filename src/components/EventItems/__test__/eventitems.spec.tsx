import React from 'react'
import TestRenderer from 'react-test-renderer'

import EventItems from '../index'
import DeleteEvent from './fixtures/DeleteEvent.json'

beforeEach(() => {
  // mock react-timeago.TimeAgo Date.now
  Date.now = jest.fn(() => 1546300800000) // 2019-01-01T00:00:00Z
})

describe('renders correctly null event', () => {
  it('should renders correctly', () => {
    const renderer = TestRenderer.create(<EventItems />)
    expect(renderer).toMatchSnapshot()
    renderer.unmount()
  })
})

describe('renders correctly empty event', () => {
  it('should renders correctly', () => {
    const renderer = TestRenderer.create(<EventItems events={[]} />)
    expect(renderer).toMatchSnapshot()
    renderer.unmount()
  })
})

describe('renders correctly EventItems', () => {
  it('DeleteEvent should renders correctly', () => {
    const renderer = TestRenderer.create(<EventItems events={DeleteEvent as any} />)
    expect(renderer).toMatchSnapshot()
    renderer.unmount()
  })
})
