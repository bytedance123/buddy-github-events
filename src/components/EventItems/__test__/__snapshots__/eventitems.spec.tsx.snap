// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly EventItems DeleteEvent should renders correctly 1`] = `
<div>
  <div
    className="sc-bdVaJa dvjgOn"
  >
    <a
      href="https://github.com/GITHUB_USERNAME"
      rel="noopener noreferrer"
      target="_blank"
    >
      <img
        alt="@GITHUB_USERNAME"
        className="sc-bwzfXH hLcALu"
        height="32"
        src="https://avatars.githubusercontent.com/u/11111?"
        width="32"
      />
    </a>
    <p
      className="sc-htpNat pFQla"
    >
      <a
        href="https://github.com/GITHUB_USERNAME"
        rel="noopener noreferrer"
        target="_blank"
      >
        GITHUB_USERNAME
      </a>
       
      deleted branch 
      fix
       at 
      <a
        href="https://github.com/GITHUB_USERNAME/PROJECT_NAME_2"
        rel="noopener noreferrer"
        target="_blank"
      >
        GITHUB_USERNAME/PROJECT_NAME_2
      </a>
    </p>
    <time
      dateTime="2019-01-01T00:00:00.000Z"
      title="2019-01-01 00:00"
    >
      0 seconds from now
    </time>
  </div>
  <div
    className="sc-bdVaJa dvjgOn"
  >
    <a
      href="https://github.com/GITHUB_USERNAME"
      rel="noopener noreferrer"
      target="_blank"
    >
      <img
        alt="@GITHUB_USERNAME"
        className="sc-bwzfXH hLcALu"
        height="32"
        src="https://avatars.githubusercontent.com/u/11111?"
        width="32"
      />
    </a>
    <p
      className="sc-htpNat pFQla"
    >
      <a
        href="https://github.com/GITHUB_USERNAME"
        rel="noopener noreferrer"
        target="_blank"
      >
        GITHUB_USERNAME
      </a>
       
      deleted tag 
      v1.0.0
       at 
      <a
        href="https://github.com/GITHUB_USERNAME/PROJECT_NAME_1"
        rel="noopener noreferrer"
        target="_blank"
      >
        GITHUB_USERNAME/PROJECT_NAME_1
      </a>
    </p>
    <time
      dateTime="2019-01-01T00:00:00.000Z"
      title="2019-01-01 00:00"
    >
      0 seconds from now
    </time>
  </div>
</div>
`;

exports[`renders correctly empty event should renders correctly 1`] = `
<div
  className="sc-bxivhb hkEkFa"
>
  No recent activity
</div>
`;

exports[`renders correctly null event should renders correctly 1`] = `null`;
